mrmustard/lab_dev/wires.py

```diff
@@ -173,13 +173,78 @@ def bra(self) -> Wires:
         "A view of this Wires object without ket wires"
         return self._view(masked_cols=(2, 3))
 
+    def add_connected(self, other) -> Wires:
+        """
+        Returns a new ``Wires`` that contains all the wires of ``self`` and ``other``, except for all
+        the output wires of ``self`` that are also input wires of ``other``.
+
+        The returned ``Wires`` corresponds to the ``Wires`` obtained by contracting the wires in common
+        between ``self`` and ``other``.
+
+        Raises:
+            ValueError: If one or more of the output wires of ``self`` that are also input wires of
+            ``other`` have different ids that the corresponding output wire of ``other``.
+        """
+        all_modes = sorted(set(self.modes) | set(other.modes))
+
+        ob = {m: 0 for m in all_modes}
+        ib = {m: 0 for m in all_modes}
+        ok = {m: 0 for m in all_modes}
+        ik = {m: 0 for m in all_modes}
+
+        msg = "Found the same wire with different ids."
+        for m in self.input.bra.modes:
+            ib[m] = self.input.bra[m].ids[0]
+        for m in self.output.bra.modes:
+            ob[m] = self.output.bra[m].ids[0]
+        for m in other.input.bra.modes:
+            if ob[m] == 0:
+                ib[m] = other.input.bra[m].ids[0]
+            elif ob[m] == other.input.bra[m].ids[0]:
+                ob[m] = 0
+            else:
+                raise ValueError(msg)
+        for m in other.output.bra.modes:
+            if ob[m] == 0:
+                ob[m] = other.output.bra[m].ids[0]
+            else:
+                raise ValueError(msg)
+
+        for m in self.input.ket.modes:
+            ik[m] = self.input.ket[m].ids[0]
+        for m in self.output.ket.modes:
+            ok[m] = self.output.ket[m].ids[0]
+        for m in other.input.ket.modes:
+            if ok[m] == 0:
+                ik[m] = other.input.ket[m].ids[0]
+            elif ok[m] == other.input.ket[m].ids[0]:
+                ok[m] = 0
+            else:
+                raise ValueError(msg)
+        for m in other.output.ket.modes:
+            if ok[m] == 0:
+                ok[m] = other.output.ket[m].ids[0]
+            else:
+                raise ValueError(msg)
+
+        combined_array = np.array([[ob[m], ib[m], ok[m], ik[m]] for m in all_modes])
+
+        return self._from_data(combined_array, all_modes)
+
     def adjoint(self) -> Wires:
         r"""
         The adjoint of this wires object, with new ids.
         """
         ob_modes, ib_modes, ok_modes, ik_modes = self._args()
         return Wires(ok_modes, ik_modes, ob_modes, ib_modes)
 
+    def dual(self) -> Wires:
```

Comment:
thanks for adding dual. adjoint and dual become properties and preserve the ids in my PR, but this is fine for now.